topi/python/topi/vision/ssd/multibox.py

```diff
@@ -0,0 +1,344 @@
+                            with ib.if_scope(tvm.any(force_suppress_node > 0,
+                                                     p_out[n * num_anchors * 6 + offset_l] ==
+                                                     p_out[n * num_anchors * 6 + offset_m])):
+                                # When force_suppress == True or class_id equals
+                                iou = calculate_overlap(p_out, n * num_anchors * 6 + offset_l + 2,
+                                                        n * num_anchors * 6 + offset_m + 2)
+                                with ib.if_scope(iou >= nms_threshold):
+                                    p_out[n * num_anchors * 6 + offset_m] = -1.0
+        with ib.else_scope():
+            with ib.for_range(0, p_valid_count[n], name="l") as l:
+                with ib.for_range(0, 6, name="m") as m:
+                    p_out[n * num_anchors * 6 + l * 6 + m] = \
+                        p_inter_out[n * num_anchors * 6 + l * 6 + m]
+        # Set invalid entry to be -1
+        with ib.for_range(0, num_anchors - p_valid_count[n], name="l") as l:
+            with ib.for_range(0, 6, name="m") as m:
+                p_out[n * num_anchors * 6 + (l + p_valid_count[n]) * 6 + m] = -1.0
+    return ib.get()
+
+
+@tvm.target.generic_func
+def multibox_detection(cls_prob, loc_pred, anchor, clip=True, threshold=0.01, nms_threshold=0.5,
+                       force_suppress=False, variances=(0.1, 0.1, 0.2, 0.2), nms_topk=-1):
+    """Convert multibox detection predictions.
+
+    Parameters
+    ----------
+    cls_prob : tvm.Tensor
+        Class probabilities.
+    loc_pred : tvm.Tensor
+        Location regression predictions.
+    anchor : tvm.Tensor
+        Prior anchor boxes.
+    clip : boolean
+        Whether to clip out-of-boundary boxes.
+    nms_threshold : float
+        Non-maximum suppression threshold.
+    force_suppress : boolean
+        Whether to suppress all detections regardless of class_id.
+    threshold : float
+        Threshold to be a positive prediction.
+    variances : tuple of float
+        Variances to be decoded from box regression output.
+    nms_topk : int
+        Keep maximum top k detections before nms, -1 for no limit.
+
+    Returns
+    -------
+    out : tvm.Tensor
+        3-D tensor with shape [batch_size, num_anchors, 6]
+    """
+    batch_size = cls_prob.shape[0]
+    num_anchors = anchor.shape[1]
+    oshape = (batch_size, num_anchors, 6)
+    # Define data alignment for intermediate buffer
+    valid_count_dal = 4
+    inter_out_dal = 64
+    sort_tensor_dal = 8
+    valid_count, inter_out = \
+        tvm.extern([(batch_size,), oshape], [cls_prob, loc_pred, anchor],
+                   lambda ins, outs: transform_loc_ir(
+                       ins[0], ins[1], ins[2], outs[0], outs[1], clip, threshold, variances),
+                   dtype=["int32", "float32"], out_data_alignment=[valid_count_dal, inter_out_dal],
```

Comment:
is a specification of alignment necessary?  by default every part uses TVM alignment

Reply:
Somehow while we define multiple outputs for extern function, the data alignment will be different from TVM default. It will give some warnings if we don't specify data alignment.

Reply:
would be nice to look into what is specifically going on.

Reply:
https://discuss.tvmlang.org/t/data-alignment-mismatch-warning/42/3